Upgrading to Infinispan 14.0.8

### Description

Infinispan 14.0.8 is the latest release. When attempting an upgrade as part of #20011 there was an exception 

```
Caused by: java.lang.RuntimeException: Component org.keycloak.models.KeycloakSessionFactory is already registered
    at org.infinispan.factories.impl.BasicComponentRegistryImpl.registerComponent(BasicComponentRegistryImpl.java:368)
    at org.infinispan.factories.AbstractComponentRegistry.registerComponentInternal(AbstractComponentRegistry.java:133)
    at org.infinispan.factories.AbstractComponentRegistry.registerComponent(AbstractComponentRegistry.java:125)
    at org.infinispan.factories.AbstractComponentRegistry.registerComponent(AbstractComponentRegistry.java:117)
    at org.keycloak.models.sessions.infinispan.initializer.InfinispanCacheInitializer.initCache(InfinispanCacheInitializer.java:68)
    at org.keycloak.models.sessions.infinispan.InfinispanUserSessionProviderFactory$6.run(InfinispanUserSessionProviderFactory.java:342)
```

This issue is about to investigate it.

### Discussion

_No response_

### Motivation

To be able to upgrade to the latest version of Infinispan to benefit from the latest bug fixes. The exception shown above might prevent this.

### Details

_No response_

--- a/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/initializer/InfinispanCacheInitializer.java
+++ b/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/initializer/InfinispanCacheInitializer.java
@@ -18,7 +18,6 @@
 package org.keycloak.models.sessions.infinispan.initializer;
 
 import org.infinispan.Cache;
-import org.infinispan.commons.CacheConfigurationException;
 import org.infinispan.commons.CacheException;
 import org.infinispan.factories.ComponentRegistry;
 import org.infinispan.manager.ClusterExecutor;
@@ -63,12 +62,12 @@ public class InfinispanCacheInitializer extends BaseCacheInitializer {
 
     @Override
     public void initCache() {
-        final ComponentRegistry cr = this.workCache.getAdvancedCache().getComponentRegistry();
-        try {
-            cr.registerComponent(sessionFactory, KeycloakSessionFactory.class);
-        } catch (UnsupportedOperationException | CacheConfigurationException ex) {
+        // due to lazy initialization, this might be called from multiple threads simultaneously, therefore, synchronize
+        synchronized (workCache) {
+            final ComponentRegistry cr = this.workCache.getAdvancedCache().getComponentRegistry();
+            // first check if already set, as Infinispan would otherwise throw a RuntimeException
             if (cr.getComponent(KeycloakSessionFactory.class) != sessionFactory) {
-                throw ex;
+                cr.registerComponent(sessionFactory, KeycloakSessionFactory.class);
             }
         }
     }

--- a/pom.xml
+++ b/pom.xml
@@ -81,8 +81,8 @@
         <jakarta.persistence.version>2.2.3</jakarta.persistence.version>
         <hibernate-orm.version>6.2.1.Final</hibernate-orm.version>
         <hibernate.c3p0.version>${hibernate-orm.version}</hibernate.c3p0.version>
-        <infinispan.version>14.0.7.Final</infinispan.version>
-        <infinispan.protostream.processor.version>4.6.0.Final</infinispan.protostream.processor.version>
+        <infinispan.version>14.0.8.Final</infinispan.version>
+        <infinispan.protostream.processor.version>4.6.2.Final</infinispan.protostream.processor.version>
         <jackson.version>2.14.2</jackson.version>
         <jackson.databind.version>2.14.2</jackson.databind.version>
         <jackson.annotations.version>${jackson.version}</jackson.annotations.version>

--- a/quarkus/runtime/src/main/resources/cache-ispn.xml
+++ b/quarkus/runtime/src/main/resources/cache-ispn.xml
@@ -18,8 +18,8 @@
 
 <infinispan
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-        xsi:schemaLocation="urn:infinispan:config:11.0 http://www.infinispan.org/schemas/infinispan-config-11.0.xsd"
-        xmlns="urn:infinispan:config:11.0">
+        xsi:schemaLocation="urn:infinispan:config:14.0 http://www.infinispan.org/schemas/infinispan-config-14.0.xsd"
+        xmlns="urn:infinispan:config:14.0">
 
     <cache-container name="keycloak">
         <transport lock-timeout="60000"/>

--- a/quarkus/runtime/src/main/resources/cache-local.xml
+++ b/quarkus/runtime/src/main/resources/cache-local.xml
@@ -18,8 +18,8 @@
 
 <infinispan
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-        xsi:schemaLocation="urn:infinispan:config:11.0 http://www.infinispan.org/schemas/infinispan-config-11.0.xsd"
-        xmlns="urn:infinispan:config:11.0">
+        xsi:schemaLocation="urn:infinispan:config:14.0 http://www.infinispan.org/schemas/infinispan-config-14.0.xsd"
+        xmlns="urn:infinispan:config:14.0">
 
     <cache-container name="keycloak">
         <local-cache name="default">

